File: 0000-0999/0085.MaximalRectangle.go
func maximalRectangle(matrix [][]byte) int {
    if len(matrix) == 0 || len(matrix[0]) == 0 {
        return 0
    }
    m, n := len(matrix), len(matrix[0])
    left, right, height := make([]int, n), make([]int, n), make([]int, n)
    for i := range right {
        right[i] = n
    }
    ans := 0
    for i := 0; i < m; i++ {
        idx_l, idx_r := 0, n
        for j := 0; j < n; j++ {
            if matrix[i][j] == '1' {
                height[j]++
                left[j] = max(left[j], idx_l)
            } else {
                height[j] = 0
                left[j], idx_l = 0, j+1
            }
        }
        for j := n-1; j >= 0; j-- {
            if matrix[i][j] == '1' {
                right[j] = min(right[j], idx_r)
            } else {
                right[j], idx_r = n, j
            }
            ans = max(ans, (right[j]-left[j])*height[j])
        }
    }
    return ans
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

File: 0000-0999/0085.MaximalRectangle.py
class Solution(object):
    def maximalRectangle(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        if not matrix:
            return 0
        m = len(matrix)
        n = len(matrix[0])
        maxRe = 0
        left, right = [0] * n, [n] * n
        height = [0] * n
        for i in xrange(m):
            cur_left, cur_right = 0, n
            for j in xrange(n):
                if matrix[i][j] == '1':
                    height[j] += 1
                    left[j] = max(left[j], cur_left)
                else:
                    height[j] = 0
                    left[j] = 0
                    cur_left = j+1
            for j in xrange(n-1, -1, -1):
                if matrix[i][j] == '1':
                    right[j] = min(right[j], cur_right)
                else:
                    right[j] = n
                    cur_right = j
            for j in xrange(n):
                maxRe = max(maxRe, (right[j]-left[j])*height[j])
        return maxRe
